File: backend/app/domain/validation/engine.py
from __future__ import annotations

from app.domain.models import GeneratedCopy, ProductBrief, Violation
from .types import Validator


def validate(
    output: GeneratedCopy,
    brief: ProductBrief,
    validators: list[Validator] | None = None,
) -> list[Violation]:
    chain = validators if validators is not None else []
    violations: list[Violation] = []
    for validator in chain:
        violations.extend(validator(output, brief))
    return violations

File: backend/src/domain/validation/engine.ts
import type { GeneratedCopy, ProductBrief, Violation } from "../models.ts";
import type { Validator } from "./types.ts";

export function validate(
  output: GeneratedCopy,
  brief: ProductBrief,
  validators: Validator[] | null = null,
): Violation[] {
  const chain = validators ?? [];
  const violations: Violation[] = [];
  for (const validator of chain) {
    violations.push(...validator(output, brief));
  }
  return violations;
}
